### ISSUE TITLE
Remapping metadata

### ISSUE THREAD

**Issue by Martii** [gerbera/gerbera#382]
Hello again.

First off, again, thanks for having this project available.

Have a question about remapping some metadata.

The usage scenario is as follows:

* ID3v2.3 *(or some other metadata structure perhaps with similar issue)*
* Have a bunch of my content using the ID3 `Long Description` tag name since it is more than 80 characters and more useful than the ID3 `Description` tag which is occasionally limited by some generators.

Is there a way in Gerbera to "remap" and serve the "Long Description" instead of serving the "Description" tag without rewriting the embedded metadata in the file? e.g. serve a modified tag set to the multi-media player? Most of my players don't care about the 80 character limitation as I've directly edited a test containers `Description` to be longer than 80 characters without issue. Most certainly a longer description shouldn't be a novel but 80 characters is usually too short to describe visually a particular content accurately in human terms.

1. I would prefer not to spend some mundane time rewriting the embedded data to truncate to 80 characters or revise it shorter. That is not an effective use of my time.
2. id3lib has the ability to pull different tag names *([Ref 1](http://id3lib.sourceforge.net/api/index.html))*... so the reiterated question is... is there a simple config entry in Gerbera to remap it via XML?
3. As far as metadata editors go... renaming a tag would probably be the same as rewriting it from past experience writing code back in my day regarding ID3. e.g. sometimes it's in a static block position vs. a zero terminated string. Even if I find a command line to export/import the tags, run a script to modify the tags... there's no guarantee that it will end up on the correct stream *(encountered this with subtitles)*. Plus I haven't found any that actually do this. They all rename via the filename to tags and not tag name to tag name.
4. Adding individual parallel files seems repetitive and mucks up folders so that is currently not an option *(e.g. .nfo files, or the like, which I'm not even sure Gerbera supports serving)*

So basically I'm left with altering an instance of Gerbera of known content that utilizes the more capable `Long Description` tag and find a way to "remap" it to the `Description` tag. This would limit this transformation to an instance and doesn't do it anywhere else that it is not needed. And of course any unknown option that someone might interject to be helpful.

Thanks for the read, helpful suggestions, and possible feature request *(Yes I read the documentation too... the `id3` XML tag doesn't appear to do this particular use case though)*

Current sys specs:
* Linux *(stable and latest)*
* macOS *(latest)*
* Windows *(stable up to unstable v10)*
* Gerbera v1.2 release
* Media players vary since I'm still testing out a lot of things with Gerbera.

**Comment by elmodaddyb** [comment]
@Martii - It looks like Gerbera does have the `"upnp:longDescription"` but it is not in use anywhere.

The system does have the capability to set the `/import/library-options/id3/auxdata` tags which will populate into the CDS Aux Data.  However, I am not familiar with its operation.

https://github.com/gerbera/gerbera/blob/master/src/metadata/taglib_handler.cc#L398

**Comment by Martii** [comment]
@elmodaddyb

> It looks like Gerbera does have the "`upnp:longDescription`"...

Something I didn't know until now. A very gracious Thank you for an awesome starting point. *(This is how I start to figure things out if I can't find that start point)*. :smiley_cat: 

> ... but it is not in use anywhere.

[Current code point][gerberaMiscUsagePoint] although no idea what this is atm. Time will give me a better understanding of this layout of this project though. :smile_cat: 

> The system does have the capability to set the `/import/library-options/id3/auxdata` tags which will populate into the CDS Aux Data.

I presume you mean [this][gerberaAuxdataForID3]. Since [this current code point][gerberaTagLibHandlerDefaults] seems to load in the "generic" or "default" metadata I can make an educated guess that the XML section is there for possible future expansion of the id3lib constants for ones that do not exist yet. I would need to drop a debug, or the normal output routine, to show me the data dump. Still haven't had time to recompile Gerbera yet though. Test platform is throwing me some curve balls that I'm not used to. 

"Nitting" on the word "set" too... it would appear that it just "add"s them. So the eventual feature request implementation could be something like a child tag of `<map-data from="something" to="somethingelse"/>` under the [`id3` section][gerberaID3section] similar to the issue that I was observing/briefly commented last month with MIME Content Type transformations. IDK yet. 


<h4>Current mappings found</h4>

| HandBrake UI        | KODI/XBMC UI |Kid3 UI           | ffmpeg      | id3lib       | Gerbera                                      | Binary Data |
|---------------------|--------------|------------------|-------------|--------------|----------------------------------------------|-------------|
| Description&#x00B9; | Plot  | Description      | description | TIT3&#x00B2; | [current code point transformation ref][gerberaRef1] | desc        |
| Plot                | ?            | Long Description | synopsis    | ?&#x00B3;    | [current code point transformation ref][gerberaRef2] | 2ldes       |


&#x00B9; - Artificially limited to 80 characters. Typically strings are zero byte terminated and not usually max length limited in ID3v2x. This fixed limitation is probably a bug or interface forced limitation possibly between container types that I will take up with this project at a later date.
&#x00B2; - Not confirmed yet but probable based off of explanation of source API at [4.2.1. Text information frames - details][id3libFrameRef]
&#x00B3; - Probably a user defined text frame with a value shown in Binary Data column.

NOTE(s):
* Confusing that no symmetry exists with nomenclature e.g. They should all probably be "Description" for human readable terms other than "tokens" for smaller sizes in code definitions and abstracts.
* If i figure out any more I'll append/edit to this legend/comment but not having much luck at the moment getting technical information from web searches. Need to dig into more available source since it is abstracted quite well.
* taglib was omitted as the application path appears to be id3lib specific but will look into that library some more in time.


[gerberaRef1]: https://github.com/gerbera/gerbera/blob/08db874d387a73d0e3ac79e49e475eeec4cd1b50/src/metadata_handler.cc#L63
[gerberaRef2]:  https://github.com/gerbera/gerbera/blob/08db874d387a73d0e3ac79e49e475eeec4cd1b50/src/metadata_handler.cc#L64

[gerberaTagLibHandlerDefaults]: https://github.com/gerbera/gerbera/blob/08db874d387a73d0e3ac79e49e475eeec4cd1b50/src/metadata/taglib_handler.cc#L159
[gerberaAuxdataForID3]: https://github.com/gerbera/gerbera/blob/08db874d387a73d0e3ac79e49e475eeec4cd1b50/doc/config-import.rst#auxdata-1
[gerberaID3section]: https://github.com/gerbera/gerbera/blob/08db874d387a73d0e3ac79e49e475eeec4cd1b50/doc/config-import.rst#id3

[gerberaMiscUsagePoint]: https://github.com/gerbera/gerbera/blob/08db874d387a73d0e3ac79e49e475eeec4cd1b50/src/atrailers_content_handler.cc#L171

[id3libFrameRef]: http://id3.org/id3v2.3.0#Text_information_frames_-_details

**Comment by elmodaddyb** [comment]
@Martii - This is a great start to a specification for the changes.  For the testing framework:

C++ - GoogleTest, you can install via `/scripts/install-googletest.sh`
UI - NodeJS + Karma Jasmine - you can run `npm install` from the `/gerbera-web` folder.

**Comment by Martii** [comment]
@elmodaddyb

Thanks for that tip on the type.

Just a thought, reiterated, perhaps a "overwrite only if not present" behavior or invert the logic... probably should keep the same behavior *(or establish one)* as the conversion from one container to the next but with a "forced" *(naming conventions to make it user friendly I think and behave like other XML entries)* switch such as:

* `<map-data from="something" to="somethingelse" overwrite="true|false"/>` e.g. if `desc` *(or it's human readable name would be better imho at a later date instead of id3lib codes mentioned at that doc)* is present do/don't overwrite it.

My test platform of Kubuntu is being a bit buggy *(Ubuntu was worse)* so it shall be a while before I can become more proficient with Gerbera here. :\ ... just throwing out the bone *(ideas)* for some bites *(discussions)* if that's okay.

**Comment by KarlStraussberger** [comment]
I just came across this issue. Most of it (mapping metadata to upnp properties) already was available in 1.8. Now I also added mapping auxdata as well.

**Comment by Martii** [comment]
OOOH! Going to have to try this out when I redo that machine. I did manage to go through my collection already but I can undo a few key ones to give it a whirl. :smiley:
